text/0000-macro-future-proofing.md

```diff
@@ -0,0 +1,147 @@
+       `T'` is `EOF` or a close delimiter of a token tree, replace `T'` with
+       `F`. If `T'` is in the set `FOLLOW(NT)`, `T'` is EOF, `T'` is any NT,
+       or `T'` is any identifier, continue. Else, reject.
+    3. Else, `T` is a complex NT.
+        1. If `T` has the form `$(...)+` or `$(...)*`, run the algorithm on
+           the contents with `F` set to `EOF`. If it accepts, continue, else,
+           reject.
+        2. If `T` has the form `$(...)U+` or `$(...)U*` for some token `U`, run
+           the algorithm on the contents with `F` set to `U`. If it accepts,
+           continue, else, reject.
+
+This algorithm should be run on every matcher in every `macro_rules`
+invocation, with `F` as `EOF`. If it rejects a matcher, an error should be
+emitted and compilation should not complete.
+
+The current legal fragment specifiers are: `item`, `block`, `stmt`, `pat`,
+`expr`, `ty`, `ident`, `path`, `meta`, and `tt`.
+
+- `FOLLOW(stmt)` = `FOLLOW(expr)`
+- `FOLLOW(pat)` = `{FatArrow, Comma, Pipe}`
+- `FOLLOW(expr)` = `{Comma, FatArrow, CloseBrace, CloseParen, CloseBracket}`
+- `FOLLOW(ty)` = `{Comma, CloseBrace, CloseParen, CloseBracket}`
+- `FOLLOW(block)` = any token
+- `FOLLOW(ident)` = any token
+- `FOLLOW(tt)` = any token
+- `FOLLOW(item)` = up for discussion
+- `FOLLOW(path)` = up for discussion
+- `FOLLOW(meta)` = up for discussion
+
+# Drawbacks
+
+It does restrict the input to a MBE, but the choice of delimiters provides
+reasonable freedom.
+
+# Alternatives
+
+1. Fix the syntax that a fragment can parse. This would create a situation
+   where a future MBE might not be able to accept certain inputs because the
+   input uses newer features than the fragment that was fixed at 1.0. For
+   example, in the `bar` MBE above, if the `ty` fragment was fixed before the
+   unboxed closure sugar was introduced, the MBE would not be able to accept
+   such a type. While this approach is feasible, it would cause unnecessary
+   confusion for future users of MBEs when they can't put certain perfectly
+   valid Rust code in the input to an MBE. Versioned fragments could avoid
+   this problem but only for new code.
+2. Keep `macro_rules` unstable. Given the great syntactical abstraction that
+   `macro_rules` provides, it would be a shame for it to be unusable in a
+   release version of Rust. If ever `macro_rules` were to be stabilized, this
+   same issue would come up.
+3. Do nothing. This is very dangerous, and has the potential to essentially
+   freeze Rust's syntax for fear of accidentally breaking a macro.
+
+# Unresolved questions
+
+1. What should the FOLLOW sets for `item`, `path`, and `meta` be?
+2. Should the `FOLLOW` set for `ty` be extended? In practice, `RArrow`,
+   `Colon`, `as`, and `in` are also used. (See next item)
+2. What, if any, identifiers should be allowed in the FOLLOW sets? The author
+   is concerned that allowing arbitrary identifiers would limit the future use
+   of "contextual keywords".
```

Comment:
If a token is in the FOLLOW set of a nt, we can never change the language in such a way that parsing that nt would consume that token. This gives us a rigid boundary around which we can change the language and not break macros.

I think the `FOLLOW` for item could be everything except `;` and `}`. I think the `FOLLOW` for `meta` *could* be anything, but I'm not sure how that interacts with future plans for letting attributes contain arbitrary token trees. I really don't know about the futures of `path`/`ty`. `path` really ought to be removed, since there isn't really a single "path" that makes sense anymore.